jax/interpreters/partial_eval.py

```diff
@@ -369,20 +387,65 @@ def instantiate_const_at(trace, instantiate, tracer):
                             ['eqn_id', 'invars', 'outvars', 'primitive',
                              'bound_subjaxprs', 'params'])
 
-def new_eqn_recipe(invars, outvars, primitive, bound_subjaxprs, params):
-  return JaxprEqnRecipe(object(), invars, map(ref, outvars), primitive,
+
+def new_eqn_recipe(invars, outvars, primitive, params,
+                   subjaxpr=None,
+                   subjaxpr_const_tracers=(),
+                   mapped_invars=()):
+  """Constructs a new JaxEqnRecipe.
+
+  Params:
+    invars: the tracers for the primitive inputs.
+    outvars: the tracers for the primitive outputs.
+    primitive: the primitive.
+    params: the primitive params
+    subjaxpr: (optional) a sub-Jaxpr, used only for `xla_call` or `xla_pmap`.
+      If present, then `subjaxpr.invars` correspond to `invars.
+      Also, `subjaxpr.constvars` correspond to `subjaxpr_const_tracers`.
+    subjaxpr_const_tracers: The constant tracers for `subjaxpr`. Must be ()
+      if not `subjaxpr`.
+    mapped_invars: for `xla_pmap` only, a list of booleans of length
+      `len(subjaxpr.invars)`, with the value True when the corresponding invar
+       is mapped over devices. If present, then `subjaxpr_const_tracers` is
+       empty (`subjaxpr` is closed).
+  """
+  if subjaxpr is not None:
+    assert len(subjaxpr.constvars) == len(subjaxpr_const_tracers)
+    assert len(subjaxpr.invars) == len(tuple(invars))
+    bound_subjaxprs = ((subjaxpr, subjaxpr_const_tracers),)
+
+    if mapped_invars:
```

Comment:
I'm not sure, but it might be simpler for the caller to add `mapped_invars` to `params`. That is, I was surprised that `JaxprTrace.process_map` wasn't doing this, but instead the logic was over here. It seems a bit weird to have an argument for `new_eqn_recipe` for `xla_pmap` only, rather than just having the pmap-specific processing code handle things.

Moreover, I'm confused about how the existing value of `params['mapped_invars']` should be handled, and I wonder if  byseparating the logic for updating `params` from the logic where we construct the new value of `mapped_invars` we're making that harder to follow. WDYT?

Reply:
I will add mapped_invars to params in the caller, it is clearer that way.